[RFC] Add experimental support for AI/RL agent integration

rouming opened the pull request:
Hi folks,

I've been working on a machine learning project called [DevilutionX-AI](https://github.com/rouming/devilutionX-AI) for a while now purely for fun and personal enjoyment. It is a reinforcement learning framework set in the Diablo environment. The framework uses `DevilutionX` with some modifications, and in this PR, I'd like to upstream these changes. Currently, this PR is marked as RFC (request for comments) because there's a possibility that something might break, tests might fail, and the tweaks I made to allow an RL agent to run `DevilutionX` instances, learn, and control the player might not be necessary for the Diablo game. However, I believe some of the fixes might be beneficial for the `DevilutionX` project, certain gameplay options might be interesting, and the approach to memory sharing in headless mode could be extremely useful for testing automation. So, my primary goal is to receive your feedback.

Moreover, with this PR, I hope to attract people who are interested in reinforcement learning (disclaimer: I don't do any RL research professionally) or those who already have experience in this field, which would be amazing. `DevilutionX-AI` is still far from completion, and the main goal of an RL agent that can clear at least the first level has not yet been achieved, so there are still many experiments and tweaks ahead.

With this PR, I'd like to ask you to take a look at the patches. I am ready to tidy up everything that aligns with the spirit of the `DevilutionX` project and anything you'd be interested in merging into the master branch. Here's a brief list of the main changes:

#### AI-Oriented Gameplay Changes

- Shared memory implementation for reinforcement learning agents. Supports external key inputs and game event monitoring. Allows a third-party application to connect to a `DevilutionX` instance, obtain the state of the game engine, and send key presses. Gives control over the player and game state monitoring.

- Enabled line-buffered stdout (`setlinebuf`) for the AI agent to consume the log from the `DevilutionX` engine. When many `DevilutionX` instances are started in parallel, the log from the engine is extremely useful for identifying abnormal situations or even seeing the backtrace of a crash.

- Added a `headless` mode option to start the game in non-windowed mode. Major mode for the AI training.

- Added an option to launch the game directly into a specified dungeon level.

- Enables deterministic level and player generation for reproducible training by setting a seed.

- Added an option to remove all monsters from the dungeon level to ease the exploration training task.

- Added an option to skip most animation ticks to accelerate training speed.

- Added an option to disable monster auto-pursuit behavior when pressing a primary action button does not lead to the pursuit of a nearby monster.

#### Various Fixes

- Fixed missing events in the main event loop when running in headless mode, which was causing the AI agent to get stuck after an event had been sent, but no reaction occurred.

- Fixed access to graphics and audio objects in `headless` mode. A few bugs were causing random crashes of the `DevilutionX` instance.

If `DevilutionX` project would benefit from adding an additional directory ["ai/"](https://github.com/rouming/devilutionX-AI/tree/master/ai), with all the Python scripts responsible for reinforcement learning training and evaluation of an agent, I would be happy to upstream this part as well.

rouming committed:
plrctrls,options: introduce Gameplay.noMonsterAutoPursuing option

This commit introduces the `Gameplay.noMonsterAutoPursuing` option in
the `diablo.ini` file, which disables monster auto-pursuing on primary
actions if set to `1`:

  Gameplay.noMonsterAutoPursuing=1

Once the primary action is pressed on a controller/pad, the hero
begins pursuing a monster. If there are a lot of monsters around,
repeatedly pressing the primary action button makes the hero run
around the entire dungeon in a completely autonomous manner.

The patch sets the amount of max steps to 0, disabling auto-pursuing.
This should be useful for teaching the AI where to move the hero
instead of always relying on the primary action button.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo,options: introduce Gameplay.gameLevel option

Introduce a new option: Gameplay.gameLevel, which lets you
load a specified level on new game start. Will be very useful
for headless mode while AI training.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo,options: fix HeadlessMode, introduce Graphics.headless option

This change introduces Graphics.headless option to run in headless
mode by changing the options ini file and not hard-code the value.

Also patch fixes HeadlessMode:

 * Don't create a window
 * Immediately start a game (exactly as for demo mode)
 * Create a new player with 'AI Agent' name, otherwise fatal error
   due to a missing save file
 * Fully disable audio/sndfx during headless mode
 * Init SDL with SDL_INIT_EVENTS, otherwise an error:
       "The event system has been shut down"

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo,options: implement shared state for reinforcement learning

Share essential game state (.bss and .data sections) over the shared
memory for reinforcement learning. Also provides two shared ring
queues for getting keys as input and share possible events.

Sharing of the game state can be enabled by the following option in
ini config:

  Share game state via file=<file>

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
options,random: introduce Gameplay.gameAndPlayerSeed

Patch introduces Gameplay.gameAndPlayerSeed for deterministic
creation of new levels. Will be needed for headless mode
and RL.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo: add more logs to print to stdout

Mostly events like pause, load, save, etc.
Will be collected by the AI worker.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
utils/log.gpp: fix annoying gcc warning

A quite legitimate warning: if FMT_EXCEPTIONS is not defined we should
return something from the function. Let it be an empty string.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo: add `setlinebuf` for stdout

diablo_state.py collects stdout prints and redirects them to stdout

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo: do not drop other events in the `ShowProgress` routine

`ShowProgress` is called on any custom event, such as when a player
warps to a new level. It is designed to handle only custom events and
discards others from the queue. This behavior results in lost
keystrokes sent by the AI agent. For instance, the issue was
reproduced when attempting to pause the game by sending the PAUSE key,
which was lost due to the behavior of `ShowProgress`.

This change introduces a poller along with the custom handler. For the
`ShowProgress` routine, the poller will peek at only custom events,
leaving others in the queue.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo,options: introduce Gameplay.noMonsters option

This change introduces `Gameplay.noMonsters` option (`Disable all
monsters=1` in `diablo.ini` file) which disables all monsters on the
level. This can be useful for AI training.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
diablo,options: introduce Gameplay.skipAnimation option

This change introduces `Gameplay.skipAnimation` option (`Skip
animation=1` in `diablo.ini` file) which disables (or at list
minimizes) ticks spent on animation. This should speed up AI
training.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

AJenbo commented:
@NiteKat competition?

AJenbo commented:
@rouming reading though your readme i would suggest making thing work coordinates relative to the player rather then absolute to the world. That makes it a lot closer to how a player interacts with the game. I'm also not sure why you expose the fully world at all time rater then just the currently visible tiles?

AJenbo commented:
By the way do you know about `--timedemo` which disables frame limitation and only runs game ticks? Might be useful for speeding up training.

glebm commented on Source/utils/mapping.cpp:
```diff
@@ -0,0 +1,212 @@
+struct entry_ptr;
+static struct entry_ptr *tail_entry;
+
+struct entry_ptr {
```
How is this meant to be used? AFAICT, it's a bunch of pointers in a memory-mapped file.
1. Is this to allow another application to get/set the values directly?
2. Why is this file memory-mapped? The things that are being pointed to don't change their address.

rouming replied:
Not quite. We place all static objects of interest in the ".shared" [section] (https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html#index-section-variable-attribute). Everything else is handled by the linker. All we need to ensure is that all objects are properly aligned in the shared file and that there are no gaps between them. This is the only purpose of the `entry_ptr` structure: every static object is registered by a pointer in a singly linked list, allowing us to traverse all the registered static objects (these pointers) during early start and check that there are no gaps (padding) in between, which is the only purpose of the `verify_no_padding()` function. 

When everything is settled and no new objects are planned for addition, the `entry_ptr` and `verify_no_padding()` function can be safely dropped. I use this verification to catch alignment bugs: when new object has added, then you access it from a python script and see some garbage, because the object was placed with 4 bytes padding.

Answering your questions:

1. No, 3rd party application does not modify anything, always reads. The only exception is the ring buffer, then 3rd party app advances the write index. But this is not related to your question.
2. Well, I assume the question would not make sense after you read the description above.

You ask valid questions. That indicates there was no description from my side about all these mechanics, sorry about that. Have to describe this carefully.

glebm replied:
I get it now, thanks for explaining.

Forcing a particular placement of static variables can impede optimization and isn't cross-platform (e.g. the mold linker doesn't support linker scripts).

Have you considered an IPC approach instead, e.g. sending events over a unix socket?
This will be more portable and also decouple internal representation from the communication protocol.

There is onging work to change the layout of some of these static structs to make them more flexible for modding, so relying on the particular layout is going to cause quite a bit of churn on your python script side.

rouming replied:
I intentionally avoided using sockets or pipes. These can introduce delays when fetching the latest state due to buffering issues. The questions about a protocol also arise: should I send everything on any small change or just send deltas? etc. While this is all doable, accessing the memory chunk directly is so straightforward that its simplicity outweighs everything else.

I am aware that this approach is not portable, but it can be hidden under a specific macro, like `SHARED_STATE`, or something. It would be disabled by default and used only when needed, which is already something.

There's also a plan B. I could eliminate these linker tricks and awkward object placement code in favor of parsing DWARF debug symbols from Python to obtain the addresses of static variables for the running `DevilutionX` instance. This would involve working with the procfs, mapping every static object from a  `/proc/$pid/mem`, instead of a single big chunk. The main challenge here is DWARF parsing, but instead of implementing this in pure Python, this could be done using extra tools (GDB, dwarfdump, etc) but I did not go into it in detail, just thoughts. The downside of this solution is (as you mentioned) is constantly broken data protocol: someone merged two static arrays in `DevilutionX`, third-party python script broke. That's why would be great to upstream everything and cover with tests :) Or I'll need to sit in a deep branch chasing the latest master HEAD.

AJenbo replied:
As long as the individual structures do not change shape won't the DWARF approach be more flexible?

rouming replied:
It will, it will. But will also require some tricks, since you can't just `mmap` a part of the external address space, as I expected from the very beginning mapping the `/proc/$pid/mem`. Sigh. I'm still looking for something as fast as just accessing memory, no IO reads, no buffering. I'll share once have working code, let's see can we move forward with this approach or not. 

I'm thinking out loud. If `DevilutionX` provides a straightforward way to access its internal engine states, then using `SWIG` or similar tools, bindings from C++/C headers for any other language could be generated. I'm personally interested in Python. I found it quite annoying to generate these bindings (mostly structures) manually, which I'm using for RL training. Do you guys think there is something practical that can be scripted and automated by accessing the internal engine state? Perhaps for tasks like unit tests or similar?

rouming commented:
> @NiteKat competition?

Awesome! Eager to see the results, @NiteKat, please share.

rouming commented:
> @rouming reading though your readme i would suggest making thing work coordinates relative to the player rather then absolute to the world. That makes it a lot closer to how a player interacts with the game. I'm also not sure why you expose the fully world at all time rater then just the currently visible tiles?

First of all, thanks for reading this through. That's a good question. Simple answer: I did that and I have the windowed approach in a separate branch. The long answer is that I'm not confident that I'm moving in the right direction with all parameters changes I make, and in order not to lose the performance already achieved ("Does the agent move to the right room? Yay! Oh no, it is stuck in the corner. Sigh" kind of a mood swing I constantly experience), I make very conservative steps forward. The windowed approach requires introducing a memory, another recurrent neural network layer. For example, RLlib, which I'm using, supports the LSTM (long short-term memory) feature, but again, I started with the simplest possible task (no monsters, full dungeon as an observation state) and am slowly increasing complexity. So of course, in the future, a windowed approach is a must, but first, I would like to see some signs of life behavior with a simpler task: "Hey agent, here is the full dungeon state! What could be easier? What will be your next action?". Another requirement in the TODO list is for the agent to make decisions based on the screenshot and pixel states, emulating a purely human-like approach, but that's another challenge.

rouming commented:
> By the way do you know about `--timedemo` which disables frame limitation and only runs game ticks? Might be useful for speeding up training.

Thanks for reminding me about that. I saw this mode in the code, and I recently did a timedemo recording for the visibility fix, so I'm aware of it but forgot to thoroughly investigate. I probably don't need the patch I created, which disables animation for player movements. But I remember you mentioned that the timedemo does not always work when certain "timed" objects are activated. Could you remind me which objects those are exactly? Don't want to debug the agent while training, this is quite time consuming.

NiteKat commented:
> > @NiteKat competition?
> 
> Awesome! Eager to see the results, @NiteKat, please share.

Hello! I don't have anything for DevilutionX specifically, but I wrote an API for use with 1.09 Diablo executable here: https://github.com/nitekat/dapi

it's not in 1.0 status but is functional enough to play a full game, both single player and multi player (all 3 difficulties). I stream an rule-based expert system in both modes regularly at my twitch. The way it works, it isn't immediately compatible with DevilutionX. We've talked about integrating the code into DevilutionX but never really worked on it.

AJenbo commented:
> But I remember you mentioned that the timedemo does not always work when certain "timed" objects are activated.

It was already addressed in the code some time ago, i just forgot.

rouming commented:
> Hello! I don't have anything for DevilutionX specifically, but I wrote an API for use with 1.09 Diablo executable here: https://github.com/nitekat/dapi
> 
> it's not in 1.0 status but is functional enough to play a full game, both single player and multi player (all 3 difficulties). I stream an rule-based expert system in both modes regularly at my twitch. The way it works, it isn't immediately compatible with DevilutionX. We've talked about integrating the code into DevilutionX but never really worked on it.

Hey! Nice, thanks for sharing. I like your windows remote process injection tricks :) Have not seen the win api for ages, cool stuff.

rouming committed:
CMake: introduce compilation test for the linker symbols

The patch includes a CMake compilation test that verifies whether the
resulting binary includes the `__bss_start` and `_end` linker
symbols. Should be the case for the GNU linker. If the binary contains
these symbols, the `HAVE_LINKER_BSS_SYMBOLS` macro will be defined.

In the next commit, the `HAVE_LINKER_BSS_SYMBOLS` macro will be used
to share the .bss and .data sections with the external
application. Stay tuned.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming commented:
> Hm the link failed for me

Strange, I picked the url from my gihub repo. It's possible that when you logged in, the url is became personalized. [This one](https://jumpshare.com/s/nfNUcRJx6JOtbRisuBqp)?

AJenbo commented:
No that's also 404, maybe it needs permissions

Diabolically-Handsome commented:
Hi @rouming — great work on this RFC, and thanks for pioneering the RL-on-DevilutionX space.

I independently built a similar project ([AlphaDiablo / DiabloGym](https://github.com/Diabolically-Handsome/AlphaDiablo)) with a different integration approach — the engine embedded in-process via pybind11 in HeadlessMode, actions injected at the network-command layer, deterministic seeded resets.

Two headless crashes we hit that your setup may also encounter once agents reach dungeon level 2+:

1. **Charging monsters** (first hit: a bat swoop): `AddRhino` → `InitMissileAnimationFromMonster` dereferences the monster's sprite list unconditionally — it's nullopt when graphics were never loaded.
2. **The Butcher's greeting**: `InitQTextMsg` → `GetSFXLength` indexes the `sgSFX` table, which is empty when audio never initializes (subtitle scroll speed wants the voice-line length).

Our patches for both are in-repo under `patches/` — happy to adapt them for this RFC, or compare notes on the headless event-pump quirks (we also hit the DisableInput handler refusing to release events, and the gamepad module braking pathfinding every tick when ControlMode stays None).

rouming commented:
> Hi @rouming — great work on this RFC, and thanks for pioneering the RL-on-DevilutionX space.

yay, glad to see more people working on this, the interest is definitely picking up.

on the patches: the rhino/missile crash I hit independently and already fixed differently in my tree, so that one is covered. the Butcher SFX crash I never hit it because the text scroll path is skipped entirely in the headless mode in my branch, but the minitext fallback in your patch looks like the right fix regardless, can be useful, thanks. my suggestion: submit those to the DevilutionX project itself.

as for where I'm at: in the develop branch (available on the github) I have an agent clearing all 16 dungeon levels, fighting monsters with melee and spells (fireball, charged bolt, stone curse, phasing), managing potions and mana shield. currently sitting at ~71% success rate averaged across all 16 levels in training. not merged to master yet but getting there.

happy to share results and stats if you're interested, also curious how your work compares at depth since the deeper levels have a very different survival problem than L1.

Diabolically-Handsome commented:
Congrats on the develop-branch milestone — per-level clears across all 16 levels with spells and mana-shield management is seriously impressive; the deep-level monster tables are no joke. Also great to see the hierarchical manager/worker split with explorer/combat options — I arrived at the same architectural instinct designing a MOBA prototype, so seeing it emerge independently here is encouraging.

Honest answer on depth: much shallower than yours, deliberately for now. Current champion is a melee Warrior living on L1: over a frozen 32-seed held-out eval (argmax-only, pinned engine commit, idle machine) it averages 35.2 kills/episode, drinks 93% of its health potions below half HP, reaches L2 in 25/32 episodes (descent ends the episode), and still dies in 12/32. The project's real product so far is the iteration ledger — fourteen generations with every failure documented. Depth is the next chapter, and your point that the deep levels are a different survival problem is exactly what I expect to relearn the hard way.

I'd gladly take you up on results/stats. Before comparing numbers I'd want to align protocols — ours are held-out frozen seeds with deterministic argmax, so I'm curious how the ~71% is measured (train-time vs held-out, sampling vs deterministic). Happy to share our eval harness in return.

And yes on the patches — I'll adapt both and submit them upstream to DevilutionX as you suggest.

Side note: I've started building [aispeedrun.ai](https://aispeedrun.ai) — a leaderboard for learned speedruns, verified by re-execution (checkpoint hash + seed + replay on a deterministic engine). Diablo I is the first category in preparation. Once the verification protocol is drafted I'd value your eyes on it — as far as I can tell, you and I are the only two people it currently concerns :)

rouming commented:
on metrics: I don't track kills or potion usage. the only thing that matters is finding the randomly placed goal on each level and surviving of course. everything else is noise for my task definition.

the ~73% is from eval during training: every 10 updates I run 250 deterministic episodes (argmax), uniformly distributed over the levels, averaged over the whole train run. fully deterministic end to end, engine seeds the world. here's the current table:

```
Per-level outcomes (8000 all eval episodes, success: 73.7% total / 73.8% uniform):
 Level        Goal       Death   Escape      Stuck  Timeout               steps(succ)               steps(fail)
---------------------------------------------------------------------------------------------------------------
     1  94.1%(482)    3.7%(19)  0.0%(0)   2.1%(11)  0.0%(0)     μ=353 σ=375 [6..2131]   μ=776 σ=524 [255..2049]
     2  50.2%(257)  29.7%(152)  0.6%(3)  18.0%(92)  1.6%(8)    μ=753 σ=524 [49..2631]    μ=952 σ=729 [21..3000]
     3  67.0%(343)  19.9%(102)  0.0%(0)  12.7%(65)  0.4%(2)    μ=397 σ=405 [14..2241]    μ=844 σ=634 [79..3000]
     4  52.7%(270)  27.9%(143)  0.2%(1)  17.8%(91)  1.4%(7)    μ=471 σ=552 [14..2280]   μ=1066 σ=713 [50..3000]
     5  63.1%(323)  23.2%(119)  1.4%(7)  12.1%(62)  0.2%(1)    μ=406 σ=396 [22..2787]    μ=920 σ=655 [56..3000]
     6  81.6%(418)   14.6%(75)  0.6%(3)   3.1%(16)  0.0%(0)     μ=400 σ=369 [4..2290]   μ=699 σ=463 [101..2754]
     7  70.1%(359)  23.8%(122)  0.4%(2)   5.5%(28)  0.2%(1)    μ=463 σ=382 [14..2906]    μ=769 σ=606 [68..3000]
     8  82.4%(422)   12.9%(66)  0.6%(3)   3.9%(20)  0.2%(1)     μ=471 σ=382 [3..2061]    μ=805 σ=544 [72..3000]
     9  87.9%(450)    6.2%(32)  0.0%(0)   5.9%(30)  0.0%(0)     μ=438 σ=320 [4..1487]   μ=927 σ=546 [154..2480]
    10  66.4%(340)  24.8%(127)  0.0%(0)   8.8%(45)  0.0%(0)    μ=371 σ=335 [25..2434]    μ=701 σ=529 [26..2718]
    11  71.2%(342)   12.7%(61)  0.0%(0)  15.8%(76)  0.2%(1)    μ=316 σ=331 [23..2377]    μ=746 σ=522 [71..3000]
    12  76.9%(369)   11.9%(57)  0.4%(2)  10.8%(52)  0.0%(0)    μ=513 σ=330 [20..1800]    μ=752 σ=473 [81..2576]
    13  83.3%(400)   15.8%(76)  0.4%(2)    0.4%(2)  0.0%(0)     μ=348 σ=339 [9..2360]   μ=541 σ=405 [142..1804]
    14  76.0%(365)   16.2%(78)  0.2%(1)   7.3%(35)  0.2%(1)     μ=631 σ=572 [8..2515]  μ=1005 σ=736 [102..3000]
    15  69.8%(335)  28.3%(136)  0.6%(3)    1.2%(6)  0.0%(0)     μ=333 σ=236 [4..1666]    μ=483 σ=260 [88..1856]
    16  88.1%(423)    9.8%(47)  0.2%(1)    1.9%(9)  0.0%(0)    μ=259 σ=259 [15..1960]    μ=612 σ=667 [54..2766]
```

the "goal" is the success rate, when goal is reached, the "death", "escape", "stuck" and "timeout" is the buckets for failures. the "steps(succ)" and "steps(fail)" is stats (mean, std, min, max) for steps when goal is reached and when is not.

> And yes on the patches — I'll adapt both and submit them upstream to DevilutionX as you suggest.

great!

> Side note: I've started building aispeedrun.ai — a leaderboard for learned speedruns, verified by re-execution (checkpoint hash + seed + replay on a deterministic engine).

ha, good to know there is something going on there, good luck with it :)

Diabolically-Handsome commented:
Thanks for the table — that failure taxonomy (death/escape/stuck/timeout) is exactly the kind of bucket split worth stealing. Two things jump out: L2 as the death valley (50.2% goal, 29.7% death — is that the Butcher's doing?), and steps(fail) sitting at ~2× steps(succ) — your failures are slow grinds, not quick deaths. One protocol question: if the 8000 episodes pool every 10-update eval across the whole run, early checkpoints drag the average down — what does your final checkpoint score on its own? You might be underselling it.

In exchange, our axes (we optimize combat, so different science): champion does 35.2 mean kills/episode on L1, deaths 12/32, 93% potion-press discipline — but same-config reruns draw 46/37/60% on that last one: outcome level replicates, behavioural style turns out to be seed lottery. Full ledger in the repo README.

Also spotted your ActionMask enum — we're about to walk the same road for our gear key. Same engine, two sciences, converging tools :)

rouming commented:
> L2 as the death valley — is that the Butcher's doing?

Butcher is disabled in training (but not in the eval, in the table I provided The Butcher acts), but even without him L2 only goes from 50% to 60%, so L2/L3/L4 are just genuinely harder levels for my agent.
 
what I actually see there: the agent collects crowds of followers while it still has potions, then runs dry, switches to luring enemies one by one from a corner, burns through the step budget and either times out or dies by accident. and the way this behavior changes actually looks really cool :) the agent quickly transforms from a very bold and self-confident fighter into a coward fleeing from bats (usually when HP drops below 20% and there are no potions at all). does not want to die, poor guy.

> steps(fail) sitting at ~2× steps(succ) — your failures are slow grinds

exactly, out of potions, overcautious, slow death by timeout.

> early checkpoints drag the average down — what does your final checkpoint score on its own?

true, but the table I provided is internal progress tracking, not something I designed for external comparison. works well for that purpose. I just got used to those numbers and know exactly when they improve or not. 

> spotted your ActionMask enum — we're about to walk the same road

that's dead code on my side, I don't use action masking :)

Diabolically-Handsome commented:
An update from the deep-water front, with a finding that might be worth a grep on your side too.

Our agents kept refusing to farm before descending. Three knobs later — depth-progressive descent bonuses (made a level-1 stair-rusher), death cost scaled by depth (pendulum swung to "never descend", and the dead started dying *fully stocked*), a char-level/dungeon-level "power gauge" in the obs (null result) — the behavior still wouldn't emerge. So we stopped debugging the agent and audited the world. Two interlocking causes:

**1. A stat-point black hole.** The engine dutifully grants 5 stat points per level (`NextPlrLevel`), but spending them is a UI action — and our bridge never implemented it. Nineteen generations of agents leveled up and silently discarded ~83% of the growth currency; a level-up was just +2 HP and a heal. Once modeled, "farm then dive" turns out to be genuinely negative-EV in that world: bare clvl3 loses to a median L3 pack by ~4 seconds, and the only winning line at reachable levels needs the vitality points spent *plus* armor. The agents weren't failing to learn the strategy — **they were correctly pricing a broken economy.** Might be worth checking on your side: your spawn injection calibrates stats at episode start, but do in-episode level-ups ever get their points spent?

**2. Discount myopia.** γ=0.99 (credit half-life ~69 steps) annihilates deferred descent bonuses. Our rusher hit the stairs at median step 138, and 8×0.99^138 ≈ 2.0 — exactly two kills' worth. The rush wasn't a pathology; it was the closed-form optimum of the prices we set.

Your potion-runway death anatomy replicated beautifully here, with a twist: under flat bonuses our divers died dry (16/22 empty belts — your coward, but bolder); after repricing death they died with potions still in the belt (15/19). A forced-drink intervention showed why: drinking changed *how* they die, not *whether* — at level 1-2, an L2-L3 pack bursts faster than any belt heals. Bottleneck chain, in order of discovery: sampling → resources → character power.

Next run fixes the black hole (auto-spend on level-up) plus a longer γ. If the spiral still doesn't emerge, your per-level calibrated-spawn curriculum is our registered fallback — the archaeology on your `player.cpp` injection is already done on our side :)

Full ledger in the repo, as always.

rouming commented:
oh jeez your messages are so dense, took me a while to understand what you mean by "black hole" and "myopia" :sweat_smile:. 
mind decrease fanciness a bit? :)

anyway, answering on both.

> stat-point black hole

I'm not in the agent phase yet. as I said previously, my ultimate goal is to have a model, which survives on all levels, exploring. how? does not matter. I do not control that, except tuning the reward function. the exp points assignment is not an AI task, during 30 years of playing Diablo 1 people discovered a very dumb and efficient scheme for assignment for each hero class. so this is out of the AI scope, this is an algorithmic task. in my case for the headless mode I just need to introduce a few API calls. should not be difficult.

> discount myopia

I see. usually I do the opposite: by all means I try to avoid farming, because then the main goal is forgotten and policy is destroyed in dozens of updates. and yes, the reward tuning is the pain. I extensively collect statistics on all reward events, so at the end of the training I can see if there is a risk for farming. so careful with that.

wish you good luck with all your RL battles :) very interesting journey. looking forward for your updates.
